Userspace: muppetjones (#1)

muppetjones opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

This PR adds:

- My userspace
- Planck keymap
- Lily58 keymap
- Kyria keymap

My userspace and keymaps incorporate code from the following users:

- @splitkb (optimizations)
- @drashna (`wrapper.h` pattern)
- @andrewjrae (casemodes)
- @j-inc (bongo cat)

I've added the following features in my userspace and keymaps. As I clean up and optimize these features, I'll add them into the main code base.

- Etch-a-Mouse: Encoder-based pointing device movement with acceleration.
- Dynamic, Layer-based RGB: RGB underglow that changes with your layers and respects your current RGB settings. (Need to move this from my kyria into my userspace).

This PR only includes one change outside of my user-specific code: Updates to `keyboards/lily58/lib/layer_state_reader.c`. The default Lily58 keymap did not compile without these updates.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

zvecr commented on keyboards/lily58/lib/layer_state_reader.c:
```diff
@@ -4,14 +4,16 @@
-#define L_LOWER (1 << 1)
-#define L_RAISE (1 << 2)
-#define L_ADJUST (1 << 3)
+#define L_LOWER (1 << 2)
```
These changes would break the current assumptions for the default keymaps and any user ones based off it.

muppetjones replied:
Rolled these back.

zvecr commented on users/muppetjones/rules.mk:
```diff
@@ -0,0 +1,17 @@
+# also requires in config.h
+#	NO_ACTION_MACRO
+#	NO_ACTION_FUNCTION
+EXTRAFLAGS += -flto
```
You want to be using the "official" way, here and elsewhere, instead
```suggestion
LTO_ENABLE = yes
```

drashna commented on keyboards/kyria/keymaps/muppetjones/rules.mk:
```diff
@@ -0,0 +1,14 @@
+
+# Reduce firmware size
+# 	https://thomasbaart.nl/2018/12/01/reducing-firmware-size-in-qmk/
+# also requires in config.h
+#	NO_ACTION_MACRO
+#	NO_ACTION_FUNCTION
+# EXTRAFLAGS += -flto
```
```suggestion
```

`LTO_ENABLE = yes` does all of this, actually.

drashna commented on keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c:
```diff
@@ -0,0 +1,135 @@
+
+#include QMK_KEYBOARD_H
+#include <stdio.h>
+#include "config.h"  // for OLED_TIMEOUT
```
This isn't needed, it is automatically added. 
```suggestion
```

drashna commented on keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c:
```diff
@@ -0,0 +1,135 @@
+/* Functions */
+
+const char *wpm_state(void) {
+  sprintf(wpm_str, "WPM: %03d", get_current_wpm());
```
This will return the value as a string: 
```suggestion
    uint8_t n = get_current_wpm();
    char wpm_counter[4];
    wpm_counter[3] = '\0';
    wpm_counter[2] = '0' + n % 10;
    wpm_counter[1] = ( n /= 10) % 10 ? '0' + (n) % 10 : (n / 10) % 10 ? '0' : ' ';
    wpm_counter[0] = n / 10 ? '0' + n / 10 : ' ';
```

And will use ~1.5kb less

muppetjones replied:
Nice! Thanks!

drashna commented on keyboards/lily58/keymaps/muppetjones/keymap.c:
```diff
@@ -0,0 +1,245 @@
+#include "muppetjones.h"
+#define LAYOUT_wrapper(...) LAYOUT(__VA_ARGS__)
+
+#ifdef PROTOCOL_LUFA
+#  include "lufa.h"
+#  include "split_util.h"
+#endif
+#ifdef SSD1306OLED
+#  include "ssd1306.h"
+#endif
```
```suggestion
```

drashna commented on keyboards/lily58/keymaps/muppetjones/keymap.c:
```diff
@@ -0,0 +1,245 @@
+#define HR_LBRC LCTL_T(KC_LBRC)
+#define HR_RBRC LSFT_T(KC_RBRC)
+
+extern uint8_t is_master;
```
```suggestion
```

drashna commented on keyboards/lily58/keymaps/muppetjones/rules.mk:
```diff
@@ -0,0 +1,42 @@
+# Reduce firmware size
+# 	https://thomasbaart.nl/2018/12/01/reducing-firmware-size-in-qmk/
+# also requires in config.h
+#	NO_ACTION_MACRO
+#	NO_ACTION_FUNCTION
+EXTRAFLAGS += -flto
```
```suggestion
LTO_ENABLE = yes
```

drashna commented on keyboards/lily58/keymaps/muppetjones/config.h:
```diff
@@ -0,0 +1,71 @@
+
+#undef RGBLED_NUM
+
+// Reduce firmware size
+// 	https://thomasbaart.nl/2018/12/01/reducing-firmware-size-in-qmk/
+// also requires EXTRAFLAGS in config.h
+#define NO_ACTION_MACRO
+#define NO_ACTION_FUNCTION
```
```suggestion
```

drashna commented on users/muppetjones/config.h:
```diff
@@ -0,0 +1,34 @@
+#pragma once
```
Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

drashna commented on users/muppetjones/features/etchamouse.h:
```diff
@@ -0,0 +1,61 @@
+
+#pragma once
+
+#ifdef POINTING_DEVICE_ENABLE
+#    ifdef ENCODER_ENABLE
```
```suggestion
#if defined(POINTING_DEVICE_ENABLE) && defined(ENCODER_ENABLE)
```

drashna commented on users/muppetjones/features/etchamouse.h:
```diff
@@ -0,0 +1,61 @@
+ */
+bool encoder_update_mouse(uint8_t index, bool clockwise);
+
+#    endif
+#endif
```
```suggestion
#endif
```

muppetjones commented on keyboards/lily58/lib/layer_state_reader.c:
```diff
@@ -7,29 +7,30 @@
+    }
 
-  return layer_state_str;
+    return layer_state_str;
```
Due to `clang-format`.

muppetjones replied:
Restored the original.

muppetjones commented:
@drashna @zvecr Made suggested changes. Thanks for the review!

drashna commented on keyboards/lily58/keymaps/muppetjones/config.h:
```diff
@@ -0,0 +1,66 @@
+  #define OLED_TIMEOUT 60000
+#endif
+
+#define USE_SERIAL_PD2
```
Not used. 
```suggestion
```

drashna commented on keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c:
```diff
@@ -0,0 +1,139 @@
+*/
+
+#include QMK_KEYBOARD_H
+#include <stdio.h>
```
```suggestion
```

drashna commented on keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c:
```diff
@@ -0,0 +1,139 @@
+// uint8_t current_prep_frame = 0; // uncomment if PREP_FRAMES >1
+uint8_t current_tap_frame = 0;
+
+char wpm_str[12];
```
```suggestion
```

drashna commented on keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c:
```diff
@@ -0,0 +1,139 @@
+    wpm_counter[2] = '0' + n % 10;
+    wpm_counter[1] = ( n /= 10) % 10 ? '0' + (n) % 10 : (n / 10) % 10 ? '0' : ' ';
+    wpm_counter[0] = n / 10 ? '0' + n / 10 : ' ';
+  return wpm_str;
```
```suggestion
    return wpm_counter;
```

muppetjones replied:
I actually just updated this last night while updating my lily58 keymap!

muppetjones commented:
@drashna Thank you for the second review. All of the comments have been addressed.

I updated and changed some things while trying to debug a [ghosting problem](https://www.reddit.com/r/olkb/comments/ofmzgz/need_help_with_ghosting_problem/?). I saw the behaviour on my kyria initially, but also when I flashed my lily with the original updates I posted here. The keys ghosting were different on the lily; however, it was still multiple keys at once. The issue seems to be in QMK as the WPM display was always high -- hanging around 128-255.

I was originally using 0.11.* with no issue, but I had pulled in the latest master (0.13.*) to get my fork up to speed. Rolling back the QMK version and the chibios version to the most recent 0.11 fixed both the ghosting and the WPM.

All of my changes actually line up with the second round of requested changes. I haven't retested on 0.13 yet. I'm going to get my testing and dev setup a little more stable and see if I can't track down the source of the above issue. If I find anything, I'll open a new PR.

drashna commented:
Yeah, there is an issue with the WPM code right now.  IIRC, there is a proposed fix for that. 

Also, develop seems to handle the split code better.

fauxpark commented on keyboards/lily58/keymaps/muppetjones/keymap.c:
```diff
@@ -0,0 +1,239 @@
+#include "muppetjones.h"
+#include "features/bongo_cat.h"
+#define LAYOUT_wrapper(...) LAYOUT(__VA_ARGS__)
+
+#ifdef PROTOCOL_LUFA
+#    include "split_util.h"
+#endif
```
Fairly sure this isn't needed.
```suggestion
```

muppetjones replied:
If I recall, it didn't compile without it.

fauxpark commented on keyboards/planck/keymaps/muppetjones/config.h:
```diff
@@ -0,0 +1,49 @@
+ * MIDI options
+ */
+
+/* Prevent use of disabled MIDI features in the keymap */
+//#define MIDI_ENABLE_STRICT 1
+
```
```suggestion
```

muppetjones replied:
Leaving this was intentional, and as comment lines aren't included at compile time, it doesn't hurt to leave them in a personal keymap.

fauxpark replied:
This define no longer exists.

fauxpark commented on keyboards/planck/keymaps/muppetjones/config.h:
```diff
@@ -0,0 +1,49 @@
+*/
+//#define MIDI_ADVANCED
+
+/* override number of MIDI tone keycodes (each octave adds 12 keycodes and allocates 12 bytes) */
+//#define MIDI_TONE_KEYCODE_OCTAVES 2
+
```
```suggestion
```

muppetjones replied:
Leaving this was intentional, and as comment lines aren't included at compile time, it doesn't hurt to leave them in a personal keymap.

fauxpark replied:
This define no longer exists.

fauxpark commented on keyboards/lily58/keymaps/muppetjones/rules.mk:
```diff
@@ -0,0 +1,41 @@
+# Build Options
+#   change to "no" to disable the options, or define them in the Makefile in
+#   the appropriate keymap folder that will get included automatically
+
+BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = no        # Mouse keys
+EXTRAKEY_ENABLE = no        # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+NKRO_ENABLE = no            # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+MIDI_ENABLE = no            # MIDI controls
+AUDIO_ENABLE = no           # Audio output on port C6
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = no       # Enable WS2812 RGB underlight.
+SWAP_HANDS_ENABLE = no      # Enable one-hand typing
+OLED_DRIVER_ENABLE= yes     # OLED display
+WPM_ENABLE = yes
+
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
+TAP_DANCE_ENABLE = yes
+
+LTO_ENABLE = yes
```
```suggestion
WPM_ENABLE = yes
TAP_DANCE_ENABLE = yes
LTO_ENABLE = yes
```

muppetjones replied:
Leaving this was intentional, and as comment lines aren't included at compile time, it doesn't hurt to leave them in a personal keymap.

This is also from the default lily58 keymap.

fauxpark replied:
The idea here is to reduce duplication. Since most of this stuff is the same as the keyboard-level rules.mk, it can be removed, leaving the keymap-level rules.mk containing only the changes vs. keyboard-level.

fauxpark commented on keyboards/lily58/keymaps/muppetjones/rules.mk:
```diff
@@ -0,0 +1,41 @@
+        # ./lib/mode_icon_reader.c \
+        # ./lib/host_led_state_reader.c \
+        # ./lib/timelogger.c \
+
+# Reduce firmware size
+# 	https://thomasbaart.nl/2018/12/01/reducing-firmware-size-in-qmk/
+# also requires in config.h
+# LTO_ENABLE = yes
```
```suggestion
```

fauxpark commented on keyboards/planck/keymaps/muppetjones/keymap.c:
```diff
@@ -0,0 +1,441 @@
+float plover_gb_song[][2] = SONG(PLOVER_GOODBYE_SOUND);
+#endif
+
+/* Setup layer lighting
+#define HSV_WHITE 0, 0, 255
+#define HSV_RED 0, 255, 255
+#define HSV_CORAL 11, 176, 255
+#define HSV_ORANGE 28, 255, 255
+#define HSV_GOLDENROD 30, 218, 218
+#define HSV_GOLD 36, 255, 255
+#define HSV_YELLOW 43, 255, 255
+#define HSV_CHARTREUSE 64, 255, 255
+#define HSV_GREEN 85, 255, 255
+#define HSV_SPRINGGREEN 106, 255, 255
+#define HSV_TURQUOISE 123, 90, 112
+#define HSV_TEAL 128, 255, 128
+#define HSV_CYAN 128, 255, 255
+#define HSV_AZURE 132, 102, 255
+#define HSV_BLUE 170, 255, 255
+#define HSV_PURPLE 191, 255, 255
+#define HSV_MAGENTA 213, 255, 255
+#define HSV_PINK 234, 128, 255
+#define HSV_BLACK 0, 0, 0
+#define HSV_OFF HSV_BLACK
+ *
+ */
```
```suggestion
```

muppetjones replied:
Leaving this was intentional, and as comment lines aren't included at compile time, it doesn't hurt to leave them in a personal keymap.

fauxpark replied:
Why duplicate what the docs already provide?

muppetjones replied:
I copied from there. I was experimenting, and wanted a more direct reference so I didn't have to look it up each time. I figured that since it was a comment, it wasn't hurting anything.

This experiment culminated in a dynamic layer lighting that I haven't moved into my user space yet.

muppetjones commented:
Not going to lie, I'm not a fan of requested changes that have zero effect on the code base (and one that actually affects compile negatively). All but one (the colors in the planck) are sections from the default keymap for that board that I had left for one reason or another. Had these been changes submitted that hit the actual codebase, I certainly would have cleaned those up beforehand, and could understand the requested changes if I had missed them.

I'm all for constructive criticism, but removing inconsequential comments and "requiring changes"...it just slows down and frustrates the review process. I've been on the reviewer side in that situation several times. A word of advice: Just "comment" instead of "request changes". If it really is critical to have the extra comments removed, please say so.

fauxpark commented:
@muppetjones it's preferable to get things like this done *before* the PR is merged and is potentially copy-pasted into other users' code, bloating the repo. It's then harder for us to clean it up afterwards, because we generally require signoff from the affected users (or at least provide a heads up) when modifying their code.

muppetjones commented:
@fauxpark That's good reasoning. I appreciate the response. I'll clean it up, and I'll see about another PR to clean up the defaults that I copied from.

muppetjones commented:
@fauxpark @zvecr I've made the requested changes -- thanks for the patience and the feedback.

muppetjones commented:
Bump for re-review when you get a chance. Thanks!
@fauxpark @zvecr

fauxpark commented on keyboards/lily58/keymaps/muppetjones/rules.mk:
```diff
@@ -0,0 +1,20 @@
+# Build Options
+#   change to "no" to disable the options, or define them in the Makefile in
+#   the appropriate keymap folder that will get included automatically
+
```
```suggestion
```

fauxpark reviewed (changes requested):
You have submodule changes.

muppetjones commented:
@fauxpark Fixed.

muppetjones commented:
Bump @fauxpark @zvecr

muppetjones commented:
@drashna Thanks for the re-review and changes. I'd kinda given up on this PR =]

I'll take a look and try to get this updated and cleaned up this weekend.

muppetjones commented on keyboards/planck/keymaps/muppetjones/config.h:
```diff
@@ -0,0 +1,44 @@
+#pragma once
```
```suggestion
 /* Copyright 2020 Stephen Bush <muppetjones>
  * 
  * This program is free software: you can redistribute it and/or modify 
  * it under the terms of the GNU General Public License as published by 
  * the Free Software Foundation, either version 2 of the License, or 
  * (at your option) any later version. 
  * 
  * This program is distributed in the hope that it will be useful, 
  * but WITHOUT ANY WARRANTY; without even the implied warranty of 
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
  * GNU General Public License for more details. 
  */
#pragma once
```

muppetjones commented:
@drashna Incorporated your suggestions. Thank you very much for the review!

muppetjones commented on keyboards/lily58/keymaps/muppetjones/README.md:
````diff
@@ -0,0 +1,13 @@
+```
+
+```
+make lily58/rev1:muppetjones:avrdude-split-left
````
```suggestion
qmk flash -kb lily58/rev1 -km muppetjones -bl avrdude-split-left
```

muppetjones commented on users/muppetjones/README.md:
```diff
@@ -0,0 +1,23 @@
+# License
```
Yep!

muppetjones replied:
Done.

drashna commented on keyboards/lily58/keymaps/muppetjones/config.h:
```diff
@@ -0,0 +1,44 @@
+/* Select hand configuration */
+
+// 1 minute
+#ifdef OLED_DRIVER_ENABLE
```
```suggestion
#ifdef OLED_ENABLE
```

muppetjones replied:
I feel like this has been changed 20 times -- I'm not sure why it keeps popping back up.

muppetjones commented:
@drashna g2g

drashna commented on keyboards/planck/keymaps/muppetjones/keymap.c:
```diff
@@ -0,0 +1,403 @@
+    return true;
+}
+
+void dip_switch_update_user(uint8_t index, bool active) {
```
Still needs to be addressed

keyboard-magpie commented:
mis-requested review sorry drashna!

muppetjones commented:
Thanks!!
